File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2019-2c-parcial2-masked/20260921-202808-59615/workdir/src/main/java/surveycrawler/GuardedRun.java
package surveycrawler;

import java.util.ArrayList;
import java.util.List;

public class GuardedRun extends CrawlerRun {

    private final List<Runnable> undoActions = new ArrayList<>();

    // guarded run

    @Override
    public void startGuardedRun(SurveyCrawler aCrawler) {
        aCrawler.signalCanNotStartGuardedRunInsideGuardedRun();
    }

    @Override
    public void endGuardedRun(SurveyCrawler aCrawler) {
        aCrawler.finishGuardedRun();
    }

    @Override
    public void assertIsFinished(SurveyCrawler aCrawler) {
        aCrawler.signalUnfinishedGuardedRun();
    }

    // moving

    @Override
    public void slideOnSilt(SurveyCrawler aCrawler, Point aDirection, SiltSlide aSlide) {
        aCrawler.signalCanNotMoveOnSiltDuringGuardedRun();
    }

    // undoing

    @Override
    public void registerUndoAction(Runnable anUndoAction) {
        undoActions.add(anUndoAction);
    }

    @Override
    public void handleErrorOf(SurveyCrawler aCrawler) {
        undoMovements();
        aCrawler.finishGuardedRun();
    }

    private void undoMovements() {
        for (int i = undoActions.size() - 1; i >= 0; i--) {
            undoActions.get(i).run();
        }
        undoActions.clear();
    }
}

File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2019-2c-parcial2-masked/20260921-202806-59410/workdir/src/surveycrawler/guarded_run.py
from surveycrawler.crawler_run import CrawlerRun


class GuardedRun(CrawlerRun):

    # initialization

    def __init__(self):
        self._undo_actions = []

    # guarded run

    def start_guarded_run(self, a_crawler):
        a_crawler.signal_can_not_start_a_guarded_run_during_a_guarded_run()

    def end_guarded_run(self, a_crawler):
        a_crawler.begin_normal_run()

    def finish(self, a_crawler):
        a_crawler.signal_guarded_run_not_finished()

    # undoing

    def register_undo(self, an_undo_action):
        self._undo_actions.append(an_undo_action)

    def handle_error(self, a_crawler):
        a_crawler.begin_normal_run()
        self.undo_all()

    def undo_all(self):
        for an_undo_action in reversed(self._undo_actions):
            an_undo_action()
        self._undo_actions = []

    # moving

    def slide_crawler_towards(self, a_crawler, a_direction, a_silt):
        a_crawler.signal_can_not_move_on_silt_during_a_guarded_run()
